web/book/src/internals/modules.md

```diff
@@ -0,0 +1,324 @@
+
+let great_tracks = my_playlists.bangers
+
+let movie_scores = my_playlists.soundtracks.movie_albums
+```
+
+Identifiers are resolved relative to current module.
+
+```prql no-eval
+module my_playlists {
+    module soundtracks {
+        let movie_albums = (from albums | filter id == 3)
+    }
+
+    from soundtracks.movie_albums
+}
+from my_playlists.soundtracks.movie_albums
+```
+
+If an identifier cannot be resolved relative to the current module, it tries to
+resolve relative to the parent module. This is repeated, stepping up the module
+hierarchy until a match is found or root of the module structure is reached.
+
+```prql no-eval
+module my_playlists {
+    let decl_1 = ...
+
+    module soundtracks {
+        let decl_2 = ...
+    }
+
+    module upbeat_rock {
+        let decl_3 = ...
+
+        from decl_1 | join soundtracks.decl2 | join decl_3
+    }
+}
+```
+
+## Main var declaration
+
+If the final declaration in a module is a variable declaration that is named
+`main`, then the leading `let main = ` can be omitted and expressed only by the
+expression itself.
+
+```
+module my_playlists {
+    let bangers = (from tracks | take 10)
+
+    from playlists | take 10
+}
+
+let album_titles = my_playlists.main
+```
+
+When a module is referenced as a value, the `main` variable is used instead.
+This is especially useful when referring to a module which is to be compiled to
+RQ (and later SQL).
+
+```
+# last line from previous example could thus be shortened to:
+let album_titles = my_playlists
+```
+
+## File importing
+
+To include PRQL source code from other files, one can use the following syntax:
+
+```
+module my_playlists
+```
+
+Such declaration loads either `./my_playlists.prql` (a leaf module) or
+`./my_playlists/_my_playlists.prql` (a directory module) and uses its contents
+as the declarations in module `my_playlists`. If none or both of the files are
+present, a compilation error is raised.
+
+Only directory modules can contain module declarations. If a leaf module
+contains a module declaration, a compilation error is raised, suggesting the
+leaf module to be converted into a directory module. This is a step toward any
+module structure having a single "normalized" representation in the file system.
+Such normalization is desired because it restrains the possible file system
+layouts to a comprehensible and predictable layout, while not sacrificing any
+functionality.
+
+Described importing rules don't achieve this "single normalized representation"
+in full, since any leaf modules could be replaced by a directory module with
+zero submodules, without any semantic changes. Restricting directory modules to
+have at least one sub-module would not improve approachability enough to justify
+adding this restriction.
+
+For example, the following module structure is annotated with files names in
+which the modules would reside:
+
+```prql no-eval
+
+module my_project {
+    # _my_project.prql
```

Comment:
How do we decide what the root module is called? What if there were a file here named `_my_project.prql`?

Reply:
Good question. I see two options:
- hardcode this to always be `_project.prql` (or something similar),
- allow any name, which would mean it will probably be the name of the project/workspace/gibberish. This would also mean it would be possible to have multiple entry-points into your project, which is something I'm -0.5 on. If there is a good use case for it then sure, but otherwise I'd avoid it.

Reply:
I think starting with `_project.prql` would be very reasonable.

If we ever need a config file, such as `Cargo.toml` or `pyproject.toml`, then we could widen the requirements later...

Reply:
> ... or even declare configs in the `_project.prql` itself?

Though PRQL isn't a great format for structured data — if we want to enable external dependencies / etc...

Reply:
Not now, but if we settle on syntax for arrays and tuples, I think we are pretty close to typed JSON.

But this is a digression in this PR.